src/os/bluestore/BlueFS.cc

```diff
@@ -2229,6 +2258,177 @@ int64_t BlueFS::_read_random(
+    }
+
+    uint64_t increase = flush_length+(sizeof(File::WALFlush::WALLength)+sizeof(File::WALFlush::WALMarker));
+    dout(20) << fmt::format("{} adding flush {:#x}~{:#x}", __func__, flush_offset, flush_length) << dendl;
+    file->wal_flushes.push_back({flush_offset, flush_length});
+    if (flush_offset >= flush_end) {
+      dout(20) << fmt::format("{} recovering flush {:#x}~{:#x}", __func__, flush_offset, flush_length) << dendl;
+      file->fnode.wal_size += flush_length;
+      file->fnode.size += increase;
+    }
+
+    flush_offset += increase;
+  }
+
+  // if we read less it might mean corruption
+  if (flush_offset < flush_end) {
+    dout(20) << fmt::format("{} read less than expected {:#x} bytes", __func__, flush_offset) << dendl;
+  }
+  ceph_assert(flush_offset >= flush_end);
+
+  delete h;
+}
+
+int64_t BlueFS::_read_wal(
+  FileReader *h,         ///< [in] read from here
+  uint64_t off,          ///< [in] offset
+  size_t len,            ///< [in] this many bytes
+  bufferlist *outbl,     ///< [out] optional: reference the result here
+  char *out)             ///< [out] optional: or copy it here
+{
+  dout(20) << __func__ << " h " << h << " offset: 0x" 
+    << off << std::hex << "~" << len << std::hex << dendl;
+  if (outbl) {
+    outbl->clear();
+  }
+
+  int64_t ret = 0;
+
+  // WAL data is wrapped in an envelope that has a format of [length of flush, payload, file ino]
+  // wal_data_logical_offset points to the offset of the payload we are currently in.
+  uint64_t wal_data_logical_offset = 0;
+
+
+  // save previous position as buffer pos is treated difffernt on regular files
+  uint64_t previous_pos = h->buf.pos;
+
+  uint64_t remaining_len = len;
+  auto flush_iterator = h->file->wal_flushes.begin();
+  while (remaining_len > 0 && flush_iterator != h->file->wal_flushes.end()) {
+    uint64_t flush_offset = flush_iterator->offset;
+    uint64_t flush_length = flush_iterator->length;
+    dout(20) << __func__ << " flush_offset " << flush_offset << dendl;
+    dout(20) << __func__ << " flush_length " << flush_length << dendl;
+
+    if (flush_length == 0) {
+      if (remaining_len > 0) {
+        dout(5) << __func__ << " flush_length 0: reading less then required "
+          << ret << "<" << len - ret << dendl;
+      }
+      break;
+    }
+    // if we won't find offset here, go ahead
+    bool in_range = wal_data_logical_offset < off+len && wal_data_logical_offset+flush_length > off;
+    if (!in_range) {
+      if (off >= wal_data_logical_offset + flush_length) {
+        // move to next flush
+        // TODO(pere): do we check "ino" here too?
+        wal_data_logical_offset += flush_length;
+        flush_iterator++;
+        continue;
+      }
+      // we can't read more, should this happen?
+      dout(5) << __func__ << " unexpected not in range: reading less then required "
+        << ret << "<" << len - ret << dendl;
+      break;
+
+    }
+
+    flush_offset += sizeof(File::WALFlush::WALLength);
```

Comment:
you wouldn't need this increment if wal_flushes would keep actual payload offset.

Reply:
I'll reuse get_payload_offset, I might've have used += here so that it was easier to understand the flow.